llvm/lib/Transforms/IPO/AttributorAttributes.cpp

```diff
@@ -10970,6 +10970,15 @@ struct AAPotentialValuesImpl : AAPotentialValues {
       Value *NewV = getSingleValue(A, *this, getIRPosition(), Values);
       if (!NewV || NewV == &OldV)
         continue;
+      // FIXME: ConstantPointerNull represents a pointer with value 0, but it
+      // doesn’t necessarily mean a nullptr. `ptr addrspace(x) null` is not the
+      // same as `addrspacecast (ptr null to ptr addrspace(x))` if the nullptr
+      // in AS x is not zero. Therefore, we can't simply replace it.
+      if (isa<ConstantPointerNull>(NewV) && isa<ConstantExpr>(OldV)) {
+        auto &CE = cast<ConstantExpr>(OldV);
```

Comment:
dyn_cast instead of isa + null

Reply:
yeah, but here we want to check both and only use the name for the second one.

Reply:
You can use the C++17 `if (auto *CE = dyn_cast<>; CE && CE->getOpcode() != ...)`